tunix/generate/utils.py

```
@@ -469,10 +487,113 @@ def process_entry(src_keys, src_val):
     new_value = (
         reshard_fn(value, tgt_param.value.sharding) if reshard_fn else value
     )
+    atol = 1e-2
+    if not jax.numpy.allclose(
+        tgt_param.value, new_value, atol=atol
+    ):
+      logging.debug('Weights for key %s do not match, Updating weights.', tgt_param)
+      logging.debug(
+          'Previous value: %s, new value: %s',tgt_param.value.ravel()[:10], new_value.ravel()[:10])
     tgt_param.value = new_value
 
+  def _extract_layer_from_scanned_tensor(tensor, layer_idx, layer_axis):
+    """Extract a specific layer from a scanned tensor."""
+    idx = [slice(None)] * tensor.ndim
+    idx[layer_axis] = layer_idx
+    return tensor[tuple(idx)]
+
+  def _get_layer_axis_from_sharding_spec(sharding_spec):
+    """Determine which axis contains the layer dimension from sharding specification."""
+    if isinstance(sharding_spec, (list, tuple)):
+      for i, spec in enumerate(sharding_spec):
+        if spec == 'layer':
+          return i
+    return None
+
+  def _should_skip_parameter(param_key):
+    """Check if a parameter should be skipped."""
+    skip_patterns = [
+        'rng',
+    ]
+    return any(pattern in param_key for pattern in skip_patterns)
+
+  def process_entry(src_keys, src_val):
+    flat_key = '.'.join(str(k) for k in src_keys)
+
+    if flat_key not in new_src_dict:
+      # Skip RNG states that don't need mapping
+      if _should_skip_parameter(flat_key):
+        logging.debug('Skipping parameter: %s', flat_key)
+      else:
+        logging.error('!!! No mapping for source key: %s', flat_key)
```

is that a logging error or better to throw an error explictly?

---

@wenxindongwork could you please respond here?

---

I think it's okay to not throw an error here, because some parameters are skipped (e.g. rng) But agreed that this can be made less error prone.

---

I'm resubmitting the PR for the weight sync utils. (It got rolled back) 
https://github.com/google/tunix/pull/231

Could you rebase after it's merged?